File: console/src/api/modules/tracing.ts
import { request } from "../request";
import { buildAuthHeaders } from "../authHeaders";

// Types
export interface OverviewStats {
  online_users: number;
  online_user_ids: string[];
  total_users: number;
  model_distribution: ModelUsage[];
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  total_sessions: number;
  total_conversations: number;
  avg_duration_ms: number;
  top_tools: ToolUsage[];
  top_skills: SkillUsage[];
  top_mcp_tools: MCPToolUsage[];
  mcp_servers: MCPServerUsage[];
  daily_trend: DailyStats[];
}

export interface ModelUsage {
  model_name: string;
  count: number;
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
}

export interface ToolUsage {
  tool_name: string;
  count: number;
  avg_duration_ms: number;
  error_count: number;
}

export interface SkillUsage {
  skill_name: string;
  count: number;
  avg_duration_ms: number;
}

export interface MCPToolUsage {
  tool_name: string;
  mcp_server: string;
  count: number;
  avg_duration_ms: number;
  error_count: number;
}

export interface MCPServerUsage {
  server_name: string;
  tool_count: number;
  total_calls: number;
  avg_duration_ms: number;
  error_count: number;
  tools: MCPToolUsage[];
}

export interface DailyStats {
  date: string;
  total_users: number;
  active_users: number;
  total_tokens: number;
  session_count: number;
}

export interface UserStats {
  user_id: string;
  model_usage: ModelUsage[];
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  total_sessions: number;
  total_conversations: number;
  avg_duration_ms: number;
  tools_used: ToolUsage[];
  skills_used: SkillUsage[];
}

export interface UserListItem {
  user_id: string;
  total_sessions: number;
  total_conversations: number;
  total_tokens: number;
  last_active: string | null;
}

export interface TraceListItem {
  trace_id: string;
  user_id: string;
  session_id: string;
  channel: string;
  start_time: string;
  duration_ms: number | null;
  total_tokens: number;
  total_input_tokens: number;
  total_output_tokens: number;
  model_name: string | null;
  status: string;
  skills_count: number;
}

export interface SessionListItem {
  session_id: string;
  user_id: string;
  channel: string;
  total_traces: number;
  total_tokens: number;
  total_skills: number;
  first_active: string | null;
  last_active: string | null;
}

export interface SessionStats {
  session_id: string;
  user_id: string;
  channel: string;
  model_usage: ModelUsage[];
  total_tokens: number;
  input_tokens: number;
  output_tokens: number;
  total_traces: number;
  avg_duration_ms: number;
  tools_used: ToolUsage[];
  skills_used: SkillUsage[];
  mcp_tools_used: MCPToolUsage[];
  first_active: string | null;
  last_active: string | null;
}

export interface TraceDetail {
  trace: Trace;
  spans: Span[];
  llm_duration_ms: number;
  tool_duration_ms: number;
  tools_called: ToolCall[];
}

export interface Trace {
  trace_id: string;
  user_id: string;
  session_id: string;
  channel: string;
  start_time: string;
  end_time: string | null;
  duration_ms: number | null;
  model_name: string | null;
  total_input_tokens: number;
  total_output_tokens: number;
  tools_used: string[];
  skills_used: string[];
  status: string;
  error: string | null;
  user_message: string | null;
}

export interface Span {
  span_id: string;
  trace_id: string;
  name: string;
  event_type: string;
  start_time: string;
  end_time: string | null;
  duration_ms: number | null;
  model_name: string | null;
  input_tokens: number | null;
  output_tokens: number | null;
  tool_name: string | null;
  skill_name: string | null;
  tool_input: Record<string, unknown> | null;
  tool_output: string | null;
  error: string | null;
}

export interface ToolCall {
  tool_name: string;
  tool_input: Record<string, unknown> | null;
  tool_output: string | null;
  duration_ms: number | null;
  error: string | null;
}

// Timeline types for hierarchical display
export interface ToolCallInSkill {
  span_id: string;
  tool_name: string;
  mcp_server: string | null;
  start_time: string;
  end_time: string | null;
  duration_ms: number;
  status: string;
  error: string | null;
  skill_weight: number | null;
}

export interface SkillCallTimeline {
  span_id: string;
  skill_name: string;
  start_time: string;
  end_time: string | null;
  duration_ms: number;
  confidence: number;
  trigger_reason: string;
  tools: ToolCallInSkill[];
  total_tool_calls: number;
  tool_duration_ms: number;
}

export interface TimelineEvent {
  event_type: string;
  span_id: string | null;
  start_time: string;
  end_time: string | null;
  duration_ms: number;
  skill_name: string | null;
  confidence: number | null;
  trigger_reason: string | null;
  tool_name: string | null;
  mcp_server: string | null;
  skill_weight: number | null;
  model_name: string | null;
  input_tokens: number | null;
  output_tokens: number | null;
  children: TimelineEvent[];
}

export interface TraceDetailWithTimeline {
  trace: Trace;
  spans: Span[];
  timeline: TimelineEvent[];
  skill_invocations: SkillCallTimeline[];
  llm_duration_ms: number;
  tool_duration_ms: number;
  skill_duration_ms: number;
  total_skills: number;
  total_tools: number;
  total_llm_calls: number;
}

export interface UserMessageItem {
  trace_id: string;
  user_id: string;
  session_id: string;
  channel: string;
  user_message: string | null;
  input_tokens: number;
  output_tokens: number;
  model_name: string | null;
  start_time: string;
  duration_ms: number | null;
}

// API functions
export const tracingApi = {
  getOverview: async (
    startDate?: string,
    endDate?: string,
  ): Promise<OverviewStats> => {
    const params = new URLSearchParams();
    if (startDate) params.append("start_date", startDate);
    if (endDate) params.append("end_date", endDate);
    return request(`/tracing/overview?${params.toString()}`);
  },

  getUsers: async (
    page = 1,
    pageSize = 20,
    filters?: {
      user_id?: string;
      start_date?: string;
      end_date?: string;
    },
  ): Promise<{
    items: UserListItem[];
    total: number;
    page: number;
    page_size: number;
  }> => {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("page_size", pageSize.toString());
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
    }
    return request(`/tracing/users?${params.toString()}`);
  },

  getUserStats: async (
    userId: string,
    startDate?: string,
    endDate?: string,
  ): Promise<UserStats> => {
    const params = new URLSearchParams();
    if (startDate) params.append("start_date", startDate);
    if (endDate) params.append("end_date", endDate);
    const query = params.toString() ? `?${params.toString()}` : "";
    return request(`/tracing/users/${encodeURIComponent(userId)}${query}`);
  },

  getTraces: async (
    page = 1,
    pageSize = 20,
    filters?: {
      user_id?: string;
      session_id?: string;
      status?: string;
      start_date?: string;
      end_date?: string;
    },
  ): Promise<{
    items: TraceListItem[];
    total: number;
    page: number;
    page_size: number;
  }> => {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("page_size", pageSize.toString());
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
    }
    return request(`/tracing/traces?${params.toString()}`);
  },

  getTraceDetail: async (traceId: string): Promise<TraceDetail> => {
    return request(`/tracing/traces/${traceId}`);
  },

  getModelUsage: async (
    startDate?: string,
    endDate?: string,
  ): Promise<{ models: ModelUsage[] }> => {
    const params = new URLSearchParams();
    if (startDate) params.append("start_date", startDate);
    if (endDate) params.append("end_date", endDate);
    const query = params.toString() ? `?${params.toString()}` : "";
    return request(`/tracing/models${query}`);
  },

  getToolUsage: async (
    startDate?: string,
    endDate?: string,
  ): Promise<{ tools: ToolUsage[] }> => {
    const params = new URLSearchParams();
    if (startDate) params.append("start_date", startDate);
    if (endDate) params.append("end_date", endDate);
    const query = params.toString() ? `?${params.toString()}` : "";
    return request(`/tracing/tools${query}`);
  },

  getSessions: async (
    page = 1,
    pageSize = 20,
    filters?: {
      user_id?: string;
      session_id?: string;
      start_date?: string;
      end_date?: string;
    },
  ): Promise<{
    items: SessionListItem[];
    total: number;
    page: number;
    page_size: number;
  }> => {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("page_size", pageSize.toString());
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
    }
    return request(`/tracing/sessions?${params.toString()}`);
  },

  getSessionStats: async (
    sessionId: string,
    startDate?: string,
    endDate?: string,
  ): Promise<SessionStats> => {
    const params = new URLSearchParams();
    if (startDate) params.append("start_date", startDate);
    if (endDate) params.append("end_date", endDate);
    const query = params.toString() ? `?${params.toString()}` : "";
    return request(
      `/tracing/sessions/${encodeURIComponent(sessionId)}${query}`,
    );
  },

  getUserMessages: async (
    page = 1,
    pageSize = 20,
    filters?: {
      user_id?: string;
      session_id?: string;
      start_date?: string;
      end_date?: string;
      query?: string;
    },
  ): Promise<{
    items: UserMessageItem[];
    total: number;
    page: number;
    page_size: number;
  }> => {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("page_size", pageSize.toString());
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
    }
    return request(`/tracing/user-messages?${params.toString()}`);
  },

  exportUserMessages: async (
    filters?: {
      user_id?: string;
      session_id?: string;
      start_date?: string;
      end_date?: string;
      query?: string;
    },
    format: string = "xlsx",
  ): Promise<Blob> => {
    const params = new URLSearchParams();
    params.append("format", format);
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value) params.append(key, value);
      });
    }
    // Use the proper API URL and include authorization token
    const { getApiUrl } = await import("../config");
    const url = getApiUrl(`/tracing/user-messages/export?${params.toString()}`);
    const headers = new Headers(buildAuthHeaders());
    const response = await fetch(url, { headers });
    if (!response.ok) {
      // Try to parse error message from response
      let errorMessage = `Export failed: ${response.status} ${response.statusText}`;
      try {
        const errorData = await response.json();
        console.error("Export error response:", errorData);
        if (errorData.detail) {
          errorMessage = errorData.detail;
        }
      } catch {
        // Ignore JSON parse error
      }
      throw new Error(errorMessage);
    }
    return response.blob();
  },

  // Timeline with skill hierarchy
  getTraceTimeline: async (traceId: string): Promise<TraceDetailWithTimeline> => {
    return request(`/tracing/traces/${traceId}/timeline`);
  },
};

File: src/copaw/app/routers/tracing.py
# -*- coding: utf-8 -*-
"""Tracing API router.

Provides REST API endpoints for tracing analytics.
"""
from datetime import datetime, timedelta
from typing import Optional
import logging

from fastapi import APIRouter, Query, HTTPException

from ...tracing import get_trace_manager, TracingConfig
from ...tracing.models import (
    OverviewStats,
    UserStats,
    UserListItem,
    TraceListItem,
    TraceDetail,
)

router = APIRouter(prefix="/tracing", tags=["tracing"])
logger = logging.getLogger(__name__)


@router.get("/overview", response_model=OverviewStats)
async def get_overview(
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
) -> OverviewStats:
    """Get overview statistics for the dashboard.

    Args:
        start_date: Optional start date filter
        end_date: Optional end date filter

    Returns:
        Overview statistics including user counts, token usage, model distribution
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        # Tracing not initialized, return empty stats
        return OverviewStats()

    start = None
    end = None

    if start_date:
        try:
            start = datetime.strptime(start_date, "%Y-%m-%d")
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid start_date format")

    if end_date:
        try:
            end = datetime.strptime(end_date, "%Y-%m-%d")
            end = end + timedelta(days=1)  # Include the end date
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid end_date format")

    return await manager.get_overview_stats(start, end)


@router.get("/users", response_model=dict)
async def get_users(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Page size"),
    user_id: Optional[str] = Query(None, description="Filter by user ID (partial match)"),
) -> dict:
    """Get list of users with their statistics.

    Args:
        page: Page number
        page_size: Page size
        user_id: Filter by user ID

    Returns:
        Paginated list of users with stats
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        return {"items": [], "total": 0, "page": page, "page_size": page_size}

    users, total = await manager.get_users(page, page_size, user_id)
    return {
        "items": [u.model_dump() for u in users],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.get("/users/{user_id}", response_model=UserStats)
async def get_user_stats(
    user_id: str,
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
) -> UserStats:
    """Get statistics for a specific user.

    Args:
        user_id: User identifier
        start_date: Optional start date filter
        end_date: Optional end date filter

    Returns:
        User statistics
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        return UserStats(user_id=user_id)

    start = None
    end = None

    if start_date:
        try:
            start = datetime.strptime(start_date, "%Y-%m-%d")
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid start_date format")

    if end_date:
        try:
            end = datetime.strptime(end_date, "%Y-%m-%d")
            end = end + timedelta(days=1)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid end_date format")

    return await manager.get_user_stats(user_id, start, end)


@router.get("/traces", response_model=dict)
async def get_traces(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Page size"),
    user_id: Optional[str] = Query(None, description="Filter by user ID"),
    session_id: Optional[str] = Query(None, description="Filter by session ID"),
    status: Optional[str] = Query(None, description="Filter by status"),
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
) -> dict:
    """Get list of traces.

    Args:
        page: Page number
        page_size: Page size
        user_id: Filter by user ID
        session_id: Filter by session ID
        status: Filter by status (running, completed, error, cancelled)
        start_date: Start date filter
        end_date: End date filter

    Returns:
        Paginated list of traces
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        return {"items": [], "total": 0, "page": page, "page_size": page_size}

    start = None
    end = None

    if start_date:
        try:
            start = datetime.strptime(start_date, "%Y-%m-%d")
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid start_date format")

    if end_date:
        try:
            end = datetime.strptime(end_date, "%Y-%m-%d")
            end = end + timedelta(days=1)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid end_date format")

    traces, total = await manager.get_traces(
        page=page,
        page_size=page_size,
        user_id=user_id,
        session_id=session_id,
        status=status,
        start_date=start,
        end_date=end,
    )
    return {
        "items": [t.model_dump() for t in traces],
        "total": total,
        "page": page,
        "page_size": page_size,
    }


@router.get("/traces/{trace_id}", response_model=TraceDetail)
async def get_trace_detail(trace_id: str) -> TraceDetail:
    """Get detailed trace with spans.

    Args:
        trace_id: Trace identifier

    Returns:
        Trace detail with all spans

    Raises:
        HTTPException: If trace not found
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        raise HTTPException(status_code=503, detail="Tracing not available")

    detail = await manager.get_trace_detail(trace_id)
    if detail is None:
        raise HTTPException(status_code=404, detail="Trace not found")

    return detail


@router.get("/models", response_model=dict)
async def get_model_usage(
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
) -> dict:
    """Get model usage statistics.

    Args:
        start_date: Start date filter
        end_date: End date filter

    Returns:
        Model usage statistics
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        return {"models": []}

    stats = await manager.get_overview_stats(
        start=datetime.strptime(start_date, "%Y-%m-%d") if start_date else None,
        end=datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1) if end_date else None,
    )

    return {"models": [m.model_dump() for m in stats.model_distribution]}


@router.get("/tools", response_model=dict)
async def get_tool_usage(
    start_date: Optional[str] = Query(None, description="Start date (YYYY-MM-DD)"),
    end_date: Optional[str] = Query(None, description="End date (YYYY-MM-DD)"),
) -> dict:
    """Get tool usage statistics.

    Args:
        start_date: Start date filter
        end_date: End date filter

    Returns:
        Tool usage statistics
    """
    try:
        manager = get_trace_manager()
    except RuntimeError:
        return {"tools": []}

    stats = await manager.get_overview_stats(
        start=datetime.strptime(start_date, "%Y-%m-%d") if start_date else None,
        end=datetime.strptime(end_date, "%Y-%m-%d") + timedelta(days=1) if end_date else None,
    )

    return {"tools": [t.model_dump() for t in stats.top_tools]}
